src/LandIce/evaluators/LandIce_BasalFrictionCoefficient_Def.hpp

```diff
@@ -242,115 +303,130 @@ KOKKOS_INLINE_FUNCTION
+        break;
+      case FIELD_TYPE::EXPONENT_OF_FIELD_AT_NODES:
+        if(nodal)
+          muValue = std::exp(muField(cell,ipt));
+        else {
+          muValue = 0;
+          for (unsigned int node=0; node<numNodes; ++node)
+            muValue += std::exp(muField(cell,node))*BF(cell,node,ipt);
         }
-      } else {
-        if (is_given_field_param) {
-          unsigned int exp_dim = is_side_equation ? numQPs : dim;
-          for (unsigned int ipt=0; ipt<exp_dim; ++ipt) {
-            beta(cell,ipt) = 0;
-            for (unsigned int node=0; node<numNodes; ++node)
-              beta(cell,ipt) += std::exp(given_field_param(cell,node))*BF(cell,node,ipt);  
-          }
-        } else {
-          unsigned int exp_dim = is_side_equation ? numQPs : dim;
-          for (unsigned int ipt=0; ipt<exp_dim; ++ipt) {
-            beta(cell,ipt) = 0;
+        break;
+      case FIELD_TYPE::EXPONENT_OF_FIELD:
+        muValue = std::exp(muField(cell,ipt));
+        break;
+      case FIELD_TYPE::UNIFORM:
+        muValue = mu;
+        break;
+      }
+
+      if (power == 1.0)
+        beta(cell,ipt) = muValue * Nval;
+      else
+        beta(cell,ipt) = muValue * Nval * std::pow (u_norm(cell,ipt), power-1.0);
+
+      if (beta_type == REGULARIZED_COULOMB) {
+        ParamScalarT lambdaValue;
+        switch (lambda_type) {
+        case FIELD_TYPE::FIELD:
+          lambdaValue = lambdaField(cell,ipt);
+          break;
+        case FIELD_TYPE::EXPONENT_OF_FIELD_AT_NODES:
+          if(nodal)
+            lambdaValue = std::exp(lambdaField(cell,ipt));
+          else {
+            lambdaValue = 0;
             for (unsigned int node=0; node<numNodes; ++node)
-              beta(cell,ipt) += std::exp(given_field(cell,node))*BF(cell,node,ipt);
+              lambdaValue += std::exp(lambdaField(cell,node))*BF(cell,node,ipt);
           }
+          break;
+        case FIELD_TYPE::EXPONENT_OF_FIELD:
+          lambdaValue = std::exp(lambdaField(cell,ipt));
+          break;
+        case FIELD_TYPE::UNIFORM:
+          lambdaValue = lambda;
+          break;
         }
+        beta(cell,ipt) /=  std::pow ( u_norm(cell,ipt) + lambdaValue*ice_softness(cell)*std::pow(Nval,n),  power);
       }
-    break;
-
-    default:
-    break;
+    }
   }
 
   if (is_side_equation && zero_on_floating) {
-    if (is_thickness_param) {
-      for (unsigned int ipt=0; ipt<dim; ++ipt) {
-        ParamScalarT isGrounded = rho_i*thickness_param_field(cell,ipt) > -rho_w*bed_topo_field_mst(cell,ipt);
-        beta(cell,ipt) *=  isGrounded;
-      }
-    } else {
-      for (unsigned int ipt=0; ipt<dim; ++ipt) {
-        ParamScalarT isGrounded = rho_i*thickness_field(cell,ipt) > -rho_w*bed_topo_field(cell,ipt);
-        beta(cell,ipt) *=  isGrounded;
+    for (unsigned int ipt=0; ipt<dim; ++ipt) {
+      MeshScalarT isGrounded;
+      if(nodal)
+        isGrounded = rho_i*thickness_field(cell,ipt) > -rho_w*bed_topo_field(cell,ipt);
```

Comment:
I'm confused on how this would work with Fad types: isGrounded is morally a bool, but how does `op >` handle the derivatives? If they are zero, then we might be better off doing
```
if (rhoi*H>-rhow*z_b)
   beta = 0.0;
```
rather than use `op *=`, which has to do more stuff (b/c of product rule in the Fad dx).

Reply:
That's my guess to. OK I'll rewrite that as if condition.